File: tests/notifications.py
"""Completion delivery with isolated TLS webhook, MQTT, FTPS and persistent SQLite.

python3 tests/notifications.py OUTPUT_DIR (builds the ignored Rust fixture service).
"""
import copy
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
import json
import os
from pathlib import Path
import queue
import sqlite3
import ssl
import subprocess
import sys
import threading
import time
import urllib.request
from print_fixture import Rig, REPO
from printer_mqtt import until

TOKEN = 'synthetic-notification-token'


class NotificationRig(Rig):
    def launch(self):
        requests = len(self.broker.requests)
        self.process = subprocess.Popen([self.binary, '--exact', 'notifications::tests::fixture_service', '--ignored', '--nocapture'],
            cwd=self.root, env=self.env, stdout=self.log, stderr=self.log)
        until(lambda: urllib.request.urlopen(self.base+'/healthz', timeout=1).status == 200, 30)
        until(lambda: len(self.broker.requests) > requests)

    def rows(self):
        with sqlite3.connect(self.store/'orca.sqlite3') as db:
            db.row_factory = sqlite3.Row
            return [dict(row) for row in db.execute('SELECT * FROM print_notifications ORDER BY rowid')]

    def sql(self, command):
        with sqlite3.connect(self.store/'orca.sqlite3') as db: db.executescript(command)

    def start_job(self):
        self.idle(); job = self.add(); count = len(self.broker.prints)
        self.next(job); until(lambda: len(self.broker.prints) == count+1)
        return job

    def finish(self):
        self.report('RUNNING'); self.phase('printing')
        self.report('FINISH'); self.phase('awaiting_removal')

    def remove_current(self):
        self.send(dict(type='discard', expected_job=self.api()['current']['id'], cleared=True))
        assert self.api()['current'] is None


def run(output):
    output = Path(output).resolve(); output.mkdir(parents=True, exist_ok=True)
    built = subprocess.check_output(['cargo', 'test', '--lib', '--no-run', '--message-format=json'], cwd=REPO, text=True)
    binary = next(v['executable'] for line in built.splitlines() if (v := json.loads(line)).get('reason') == 'compiler-artifact' and v.get('executable') and v['target']['kind'] == ['lib'])
    rig = NotificationRig(binary, output); actions = queue.Queue(); requests = []; results = {}
    class Webhook(BaseHTTPRequestHandler):
        def do_POST(self):
            payload = json.loads(self.rfile.read(int(self.headers['Content-Length'])))
            requests.append(dict(time=time.monotonic(), path=self.path, payload=payload))
            assert self.path == f'/api/webhooks/123/{TOKEN}?wait=true'
            action = actions.get_nowait() if not actions.empty() else (200, {'id': str(1000+len(requests))}, 0)
            status, body, delay = action
            time.sleep(delay)
            data = json.dumps(body).encode()
            try:
                self.send_response(status); self.send_header('Content-Type', 'application/json')
                self.send_header('Content-Length', str(len(data))); self.end_headers(); self.wfile.write(data)
            except (BrokenPipeError, ConnectionResetError, ssl.SSLError): pass
        def log_message(self, *_): pass
    webhook = ThreadingHTTPServer(('127.0.0.1', 0), Webhook)
    subprocess.run(['openssl', 'req', '-config', '/dev/null', '-x509', '-newkey', 'ec', '-pkeyopt', 'ec_paramgen_curve:P-256',
        '-addext', 'subjectAltName=DNS:localhost', '-nodes', '-days', '1', '-subj', '/CN=localhost',
        '-keyout', str(rig.root/'webhook.key'), '-out', str(rig.root/'webhook.pem')], check=True, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
    tls = ssl.SSLContext(ssl.PROTOCOL_TLS_SERVER); tls.load_cert_chain(rig.root/'webhook.pem', rig.root/'webhook.key')
    webhook.socket = tls.wrap_socket(webhook.socket, server_side=True)
    threading.Thread(target=webhook.serve_forever, daemon=True).start()
    rig.env.pop('DISCORD_WEBHOOK_URL', None); rig.env.pop('ORCA_PUBLIC_URL', None)
    rig.env.update(NOTIFICATION_TEST_ENABLED='1', NOTIFICATION_TEST_CERT=str(rig.root/'webhook.pem'),
        NOTIFICATION_TEST_URL=f'https://localhost:{webhook.server_port}/api/webhooks/123/{TOKEN}?wait=true')
    try:
        rig.launch(); rig.seed()
        plate = rig.api('/api/plates/'+rig.plate['id']); plate['name'] = '@everyone **fixture** <@123>'
        rig.plate = rig.api('/api/plates/'+plate.pop('id'), plate, 'PUT')
        first = rig.start_job()
        command = rig.broker.prints[-1]
        rig.broker.send({'print': dict(command='project_file', sequence_id=command['sequence_id'], result='success')})
        until(lambda: rig.api()['printer']['start']['phase'] == 'accepted')
        value = copy.deepcopy(rig.full); value['print'].update(gcode_state='RUNNING', mc_percent=100, subtask_name=command['subtask_name'], gcode_file=command['file'])
        rig.broker.send(value); rig.phase('printing'); time.sleep(1.1)
        assert not requests and not rig.rows()
        rig.report('FINISH'); rig.phase('awaiting_removal')
        until(lambda: rig.rows() and rig.rows()[0]['state'] == 'sent')
        sent = rig.rows()[0]; assert sent['job_id'] == first['id'] and sent['message_id'] == '1001'
        text = requests[0]['payload']['content']
        assert '取り外し待ち' in text and first['id'] in text and 'https://orca.example/queue?printer_id=p1' in text
        assert '@everyone' not in text and '<@123>' not in text and requests[0]['payload']['allowed_mentions'] == {'parse': []}
        for _ in range(3): rig.report('FINISH')
        rig.stop(); rig.launch(); rig.report('FINISH'); rig.phase('awaiting_removal'); time.sleep(1.2)
        assert len(requests) == 1 and rig.rows()[0]['tries'] == 1
        rig.remove_current(); assert len(rig.rows()) == 1
        results['finish_only_one_notification_across_reports_restart_and_removal'] = True
        # A pending outcome survives immediate removal. Slow webhook does not hold queue/MQTT.
        actions.put((200, {'id':'2000'}, 1.2)); second = rig.start_job(); rig.finish()
        until(lambda: len(requests) == 2)
        started = time.monotonic(); rig.remove_current(); assert time.monotonic()-started < .5
        until(lambda: rig.rows()[-1]['state'] == 'unknown')
        assert rig.rows()[-1]['job_id'] == second['id']
        until(lambda: rig.rows()[-1]['state'] == 'sent')
        assert rig.rows()[-1]['tries'] == 2  # Reply loss can duplicate an accepted message.
        results['timeout_does_not_block_removal_and_retains_uncertainty'] = True
        # A rate limit persists over restart; no request before the requested delay.
        actions.put((429, {'retry_after':3.0}, 0)); rig.start_job(); rig.finish()
        until(lambda: rig.rows()[-1]['result'] == '429'); limited = requests[-1]['time']; count = len(requests)
        rig.stop(); rig.launch(); rig.report('FINISH'); time.sleep(.5); assert len(requests) == count
        until(lambda: rig.rows()[-1]['state'] == 'sent')
        assert requests[-1]['time']-limited >= 3 and rig.rows()[-1]['tries'] == 2
        rig.remove_current(); results['rate_limit_survives_restart'] = True
        for label, responses, state, tries in [
            ('server_error_recovers', [(503, {}, 0), (200, {'id':'3000'}, 0)], 'sent', 2),
            ('server_error_stops', [(503, {}, 0)]*3, 'failed', 3),
            ('permanent_error_stops', [(401, {'message':TOKEN}, 0)], 'failed', 1),
            ('lost_response_stops_uncertain', [(200, {'id':'4000'}, 1.2)]*3, 'unknown', 3),
        ]:
            for action in responses: actions.put(action)
            rig.start_job(); rig.finish()
            until(lambda: rig.rows()[-1]['state'] == state and rig.rows()[-1]['tries'] == tries, 20)
            count = len(requests); time.sleep(1.2); assert len(requests) == count
            assert rig.api()['current']['state'] == 'awaiting_removal'
            rig.remove_current(); results[label] = True
        # Process loss after the HTTP request is uncertain; bounded replay resumes from disk.
        actions.put((200, {'id':'5000'}, 1.2)); rig.start_job(); count = len(requests); rig.finish()
        until(lambda: len(requests) > count)
        assert rig.rows()[-1]['state'] == 'sending'
        rig.stop(kill=True); rig.launch(); rig.report('FINISH')
        assert rig.rows()[-1]['state'] == 'unknown'
        until(lambda: rig.rows()[-1]['state'] == 'sent')
        assert rig.rows()[-1]['tries'] == 2; rig.remove_current()
        results['kill_in_flight_recovers_with_possible_duplicate'] = True
        # Save failure after an acknowledgement retries storage, without another HTTP delivery.
        rig.sql("CREATE TRIGGER fixture_fail_ack BEFORE UPDATE ON print_notifications WHEN NEW.state='sent' BEGIN SELECT RAISE(ABORT,'fixture'); END;")
        rig.start_job(); count = len(requests); rig.finish(); until(lambda: len(requests) > count)
        time.sleep(.2); assert rig.rows()[-1]['state'] == 'sending'
        rig.sql('DROP TRIGGER fixture_fail_ack;'); count = len(requests)
        until(lambda: rig.rows()[-1]['state'] == 'sent'); assert len(requests) == count
        rig.remove_current(); results['ack_save_retry_does_not_resend'] = True
        # A rejected attempt and cancellation generate no completion; explicit retry can complete.
        count = len(requests); job = rig.start_job(); command = rig.broker.prints[-1]
        rig.broker.send({'print':dict(command='project_file', sequence_id=command['sequence_id'], result='fail')})
        rig.phase('needs_attention'); time.sleep(1.2); assert len(requests) == count
        rig.idle(); rig.send(dict(type='retry', expected_job=job['id'], cleared=True)); old = command
        until(lambda: rig.broker.prints[-1] != old); rig.finish()
        until(lambda: rig.rows()[-1]['state'] == 'sent'); assert len(requests) == count+1
        rig.remove_current(); cancelled = rig.add(); rig.send(dict(type='remove',job_id=cancelled['id']))
        time.sleep(1.2); assert len(requests) == count+1
        results['failed_and_cancelled_do_not_notify_retry_attempt_does'] = True
        # Disable then re-enable across a schema-7 migration: existing FINISH is not backfilled.
        rig.stop(); rig.env['NOTIFICATION_TEST_ENABLED'] = '0'; rig.launch()
        rig.start_job(); rig.finish(); count = len(requests); before = len(rig.rows()); time.sleep(1.2)
        assert len(requests) == count and len(rig.rows()) == before
        rig.stop(); rig.sql('ALTER TABLE plates DROP COLUMN brim_enabled; ALTER TABLE plates DROP COLUMN deleted; ALTER TABLE plates DROP COLUMN sparse_infill_pattern; ALTER TABLE plates DROP COLUMN sparse_infill_density; ALTER TABLE plates DROP COLUMN wall_loops; ALTER TABLE default_settings DROP COLUMN sparse_infill_pattern; ALTER TABLE default_settings DROP COLUMN sparse_infill_density; ALTER TABLE default_settings DROP COLUMN wall_loops; DROP TABLE print_notifications; ALTER TABLE print_jobs DROP COLUMN estimate_json; PRAGMA user_version=7;')
        rig.env['NOTIFICATION_TEST_ENABLED'] = '1'; rig.launch(); rig.report('FINISH'); rig.phase('awaiting_removal'); time.sleep(1.2)
        assert not rig.rows() and len(requests) == count
        rig.remove_current(); results['unset_disables_and_schema_upgrade_does_not_backfill'] = True
        for path in ['/api/queue?printer_id=p1','/api/printers','/api/plates','/api/about']:
            assert TOKEN not in json.dumps(rig.api(path))
        assert TOKEN not in (output/'server.log').read_text()
        results.update(secrets_redacted=True, https_verified=True, requests=len(requests), print_attempts=len(rig.broker.prints))
        (output/'result.json').write_text(json.dumps(results, indent=2)); print(json.dumps(results))
    finally:
        webhook.shutdown(); webhook.server_close(); rig.close()


if __name__ == '__main__': run(sys.argv[1])

File: tests/scenarios/notifications.rs
//! Private notification-client injection stays in cfg(test); all external peers are loopback Rust.
#[path = "../common/mod.rs"]
mod common;
use common::*;
use serde_json::{Value, json};
use std::{
    collections::VecDeque,
    fs,
    io::{BufRead, BufReader, Read, Write},
    net::TcpListener,
    process::Command,
    sync::{
        Arc, Mutex,
        atomic::{AtomicBool, Ordering},
    },
    thread::{self, JoinHandle},
    time::{Duration, Instant},
};
const TOKEN: &str = "synthetic-notification-token";
type Reply = (u16, Value, Duration);
struct Webhook {
    url: String,
    requests: Arc<Mutex<Vec<(Instant, Value)>>>,
    actions: Arc<Mutex<VecDeque<Reply>>>,
    stop: Arc<AtomicBool>,
    thread: Option<JoinHandle<()>>,
}
impl Webhook {
    fn new(root: &std::path::Path) -> Self {
        let output = Command::new("openssl")
            .args([
                "req",
                "-config",
                "/dev/null",
                "-x509",
                "-newkey",
                "ec",
                "-pkeyopt",
                "ec_paramgen_curve:P-256",
                "-addext",
                "subjectAltName=DNS:localhost",
                "-nodes",
                "-days",
                "1",
                "-subj",
                "/CN=localhost",
                "-keyout",
            ])
            .arg(root.join("webhook.key"))
            .arg("-out")
            .arg(root.join("webhook.pem"))
            .output()
            .unwrap();
        assert!(output.status.success());
        let config = common::peers::config(&root.join("webhook.pem"), &root.join("webhook.key"));
        let listener = TcpListener::bind("127.0.0.1:0").unwrap();
        listener.set_nonblocking(true).unwrap();
        let url = format!(
            "https://localhost:{}/api/webhooks/123/{TOKEN}?wait=true",
            listener.local_addr().unwrap().port()
        );
        let requests = Arc::new(Mutex::new(Vec::new()));
        let actions = Arc::new(Mutex::new(VecDeque::<Reply>::new()));
        let stop = Arc::new(AtomicBool::new(false));
        let received = requests.clone();
        let replies = actions.clone();
        let stopped = stop.clone();
        let thread = thread::spawn(move || {
            let mut workers = Vec::new();
            while !stopped.load(Ordering::SeqCst) {
                match listener.accept() {
                    Ok((raw, _)) => {
                        let config = config.clone();
                        let received = received.clone();
                        let replies = replies.clone();
                        workers.push(thread::spawn(move||{let stream=common::peers::tls(raw,&config).unwrap();let mut reader=BufReader::new(stream);let mut line=String::new();reader.read_line(&mut line).unwrap();assert_eq!(line.trim(),format!("POST /api/webhooks/123/{TOKEN}?wait=true HTTP/1.1"));let mut length=None;loop{line.clear();reader.read_line(&mut line).unwrap();if line=="\r\n"{break;}
if let Some((key,value))=line.split_once(':')&& key.eq_ignore_ascii_case("content-length"){length=Some(value.trim().parse::<usize>().unwrap());}}let length=length.unwrap();assert!(length<1024*1024);let mut body=vec![0;length];reader.read_exact(&mut body).unwrap();let payload=serde_json::from_slice(&body).unwrap();let count={let mut received=received.lock().unwrap();received.push((Instant::now(),payload));received.len()};let(status,body,delay)=replies.lock().unwrap().pop_front().unwrap_or((200,json!({"id":(1000+count).to_string()}),Duration::ZERO));thread::sleep(delay);let data=serde_json::to_vec(&body).unwrap();let header=format!("HTTP/1.1 {status} Fixture\r\nContent-Type: application/json\r\nContent-Length: {}\r\nConnection: close\r\n\r\n",data.len());let _=reader.get_mut().write_all(&[header.as_bytes(),&data].concat());}));
                    }
                    Err(e) if e.kind() == std::io::ErrorKind::WouldBlock => {
                        thread::sleep(Duration::from_millis(10));
                    }
                    Err(e) => panic!("webhook accept: {e}"),
                }
            }
            for worker in workers {
                worker.join().unwrap();
            }
        });
        Self {
            url,
            requests,
            actions,
            stop,
            thread: Some(thread),
        }
    }
    fn len(&self) -> usize {
        self.requests.lock().unwrap().len()
    }
    fn respond(&self, status: u16, body: Value, delay: f64) {
        self.actions
            .lock()
            .unwrap()
            .push_back((status, body, Duration::from_secs_f64(delay)));
    }
}
impl Drop for Webhook {
    fn drop(&mut self) {
        self.stop.store(true, Ordering::SeqCst);
        let result = self.thread.take().unwrap().join();
        if !thread::panicking() {
            result.unwrap();
        }
    }
}
fn rows(rig: &Rig) -> Vec<Value> {
    let db = rig.db();
    db.prepare("SELECT job_id,state,message_id,tries,result FROM print_notifications ORDER BY rowid").unwrap().query_map([],|r|Ok(json!({"job_id":r.get::<_,String>(0)?,"state":r.get::<_,String>(1)?,"message_id":r.get::<_,Option<String>>(2)?,"tries":r.get::<_,i64>(3)?,"result":r.get::<_,Option<String>>(4)?}))).unwrap().collect::<Result<_,_>>().unwrap()
}
fn last(rig: &Rig) -> Value {
    rows(rig).last().cloned().unwrap_or(Value::Null)
}
fn start_job(rig: &mut Rig) -> Value {
    rig.idle();
    let job = rig.add(3);
    let count = rig.broker.prints().len();
    rig.next(&job, 200);
    until(|| rig.broker.prints().len() == count + 1, 12);
    job
}
fn remove_current(rig: &Rig) {
    rig.discard();
    assert!(rig.queue()["current"].is_null());
}
fn sleep(seconds: f64) {
    thread::sleep(Duration::from_secs_f64(seconds));
}
#[test]
#[allow(clippy::too_many_lines)]
fn isolated_notification_delivery() {
    let mut rig = Rig::new("notifications");
    rig.binary = std::env::current_exe().unwrap();
    rig.arguments = [
        "--exact",
        "notifications::tests::fixture_service",
        "--ignored",
        "--nocapture",
    ]
    .map(str::to_owned)
    .to_vec();
    let webhook = Webhook::new(rig.root.path());
    rig.env
        .insert("NOTIFICATION_TEST_ENABLED".into(), "1".into());
    rig.env.insert(
        "NOTIFICATION_TEST_CERT".into(),
        rig.root.path().join("webhook.pem").display().to_string(),
    );
    rig.env
        .insert("NOTIFICATION_TEST_URL".into(), webhook.url.clone());
    rig.launch();
    rig.seed();
    let path = format!("/api/plates/{}", id(&rig.plate));
    let mut plate = edit(&rig.get(&path));
    plate["name"] = json!("@everyone **fixture** <@123>");
    rig.plate = rig.put(&path, &plate, 200);
    let first = start_job(&mut rig);
    let command = rig.broker.prints().last().unwrap().clone();
    rig.broker.send(&json!({"print":{"command":"project_file","sequence_id":command["sequence_id"],"result":"success"}}));
    rig.start_phase("accepted");
    let mut value = rig.full.clone();
    value["print"] = merge(
        &value["print"],
        &json!({"gcode_state":"RUNNING","mc_percent":100,"subtask_name":command["subtask_name"],"gcode_file":command["file"]}),
    );
    rig.broker.send(&value);
    rig.phase("printing");
    sleep(1.1);
    assert_eq!(webhook.len(), 0);
    assert!(rows(&rig).is_empty());
    let stopped = rig.queue()["current"]["attempt_id"].clone();
    rig.report("FAILED");
    rig.phase("needs_attention");
    rig.send(
        json!({"type":"retry","expected_job":first["id"],"cleared":true}),
        200,
    );
    until(|| rig.broker.prints().len() == 2, 12);
    let restarted = rig.queue()["current"]["attempt_id"].clone();
    assert_ne!(restarted, stopped);
    assert!(rows(&rig).is_empty());
    rig.report("RUNNING");
    rig.phase("printing");
    rig.report("FINISH");
    rig.phase("awaiting_removal");
    until(|| last(&rig)["state"] == "sent", 12);
    let sent = rows(&rig)[0].clone();
    assert_eq!(
        rig.db()
            .query_row("SELECT attempt_id FROM print_notifications", [], |r| r
                .get::<_, String>(
                0
            ))
            .unwrap(),
        restarted.as_str().unwrap()
    );
    assert_eq!(sent["job_id"], first["id"]);
    assert_eq!(sent["message_id"], "1001");
    let payload = webhook.requests.lock().unwrap()[0].1.clone();
    let text = payload["content"].as_str().unwrap();
    assert!(
        text.contains("取り外し待ち")
            && text.contains(id(&first))
            && text.contains("https://orca.example/queue?printer_id=p1")
    );
    assert!(!text.contains("@everyone") && !text.contains("<@123>"));
    assert_eq!(payload["allowed_mentions"], json!({"parse":[]}));
    for _ in 0..3 {
        rig.report("FINISH");
    }
    rig.stop(false);
    rig.launch();
    rig.report("FINISH");
    rig.phase("awaiting_removal");
    sleep(1.2);
    assert_eq!(webhook.len(), 1);
    assert_eq!(rows(&rig)[0]["tries"], 1);
    remove_current(&rig);
    assert_eq!(rows(&rig).len(), 1);
    webhook.respond(200, json!({"id":"2000"}), 1.2);
    let second = start_job(&mut rig);
    rig.finish();
    until(|| webhook.len() == 2, 12);
    let started = Instant::now();
    remove_current(&rig);
    assert!(started.elapsed() < Duration::from_millis(500));
    until(|| last(&rig)["state"] == "unknown", 12);
    assert_eq!(last(&rig)["job_id"], second["id"]);
    until(|| last(&rig)["state"] == "sent", 12);
    assert_eq!(last(&rig)["tries"], 2);
    webhook.respond(429, json!({"retry_after":3.0}), 0.0);
    start_job(&mut rig);
    rig.finish();
    until(|| last(&rig)["result"] == "429", 12);
    let limited = webhook.requests.lock().unwrap().last().unwrap().0;
    let count = webhook.len();
    rig.stop(false);
    rig.launch();
    rig.report("FINISH");
    sleep(0.5);
    assert_eq!(webhook.len(), count);
    until(|| last(&rig)["state"] == "sent", 12);
    assert!(
        webhook
            .requests
            .lock()
            .unwrap()
            .last()
            .unwrap()
            .0
            .duration_since(limited)
            >= Duration::from_secs(3)
    );
    assert_eq!(last(&rig)["tries"], 2);
    remove_current(&rig);
    for (responses, state, tries) in [
        (
            vec![(503, json!({}), 0.0), (200, json!({"id":"3000"}), 0.0)],
            "sent",
            2,
        ),
        (vec![(503, json!({}), 0.0); 3], "failed", 3),
        (vec![(401, json!({"message":TOKEN}), 0.0)], "failed", 1),
        (vec![(200, json!({"id":"4000"}), 1.2); 3], "unknown", 3),
    ] {
        for (status, body, delay) in responses {
            webhook.respond(status, body, delay);
        }
        start_job(&mut rig);
        rig.finish();
        until(
            || last(&rig)["state"] == state && last(&rig)["tries"] == tries,
            20,
        );
        let count = webhook.len();
        sleep(1.2);
        assert_eq!(webhook.len(), count);
        assert_eq!(rig.queue()["current"]["state"], "awaiting_removal");
        remove_current(&rig);
    }
    webhook.respond(200, json!({"id":"5000"}), 1.2);
    start_job(&mut rig);
    let count = webhook.len();
    rig.finish();
    until(|| webhook.len() > count, 12);
    assert_eq!(last(&rig)["state"], "sending");
    rig.stop(true);
    rig.launch();
    rig.report("FINISH");
    assert_eq!(last(&rig)["state"], "unknown");
    until(|| last(&rig)["state"] == "sent", 12);
    assert_eq!(last(&rig)["tries"], 2);
    remove_current(&rig);
    rig.db().execute_batch("CREATE TRIGGER fixture_fail_ack BEFORE UPDATE ON print_notifications WHEN NEW.state='sent' BEGIN SELECT RAISE(ABORT,'fixture'); END;").unwrap();
    start_job(&mut rig);
    let count = webhook.len();
    rig.finish();
    until(|| webhook.len() > count, 12);
    sleep(0.2);
    assert_eq!(last(&rig)["state"], "sending");
    rig.db()
        .execute_batch("DROP TRIGGER fixture_fail_ack;")
        .unwrap();
    let count = webhook.len();
    until(|| last(&rig)["state"] == "sent", 12);
    assert_eq!(webhook.len(), count);
    remove_current(&rig);
    let count = webhook.len();
    let job = start_job(&mut rig);
    let command = rig.broker.prints().last().unwrap().clone();
    rig.broker.send(&json!({"print":{"command":"project_file","sequence_id":command["sequence_id"],"result":"fail"}}));
    rig.phase("needs_attention");
    sleep(1.2);
    assert_eq!(webhook.len(), count);
    rig.idle();
    rig.send(
        json!({"type":"retry","expected_job":job["id"],"cleared":true}),
        200,
    );
    until(|| rig.broker.prints().last() != Some(&command), 12);
    rig.finish();
    until(|| last(&rig)["state"] == "sent", 12);
    assert_eq!(webhook.len(), count + 1);
    remove_current(&rig);
    let cancelled = rig.add(3);
    rig.send(json!({"type":"remove","job_id":cancelled["id"]}), 200);
    sleep(1.2);
    assert_eq!(webhook.len(), count + 1);
    rig.stop(false);
    rig.env
        .insert("NOTIFICATION_TEST_ENABLED".into(), "0".into());
    rig.launch();
    start_job(&mut rig);
    rig.finish();
    let count = webhook.len();
    let before = rows(&rig).len();
    sleep(1.2);
    assert_eq!(webhook.len(), count);
    assert_eq!(rows(&rig).len(), before);
    rig.stop(false);
    legacy_schema::queue_v16(&rig.db());
    rig.db().execute_batch("DROP TABLE print_history; ALTER TABLE plate_items DROP COLUMN roles_json; ALTER TABLE plates DROP COLUMN secondary_filament_id; DROP TABLE plate_imports; ALTER TABLE plates DROP COLUMN support_interface_filament_id; ALTER TABLE plates DROP COLUMN support_enabled; ALTER TABLE plates DROP COLUMN brim_enabled; ALTER TABLE plates DROP COLUMN deleted; ALTER TABLE plates DROP COLUMN sparse_infill_pattern; ALTER TABLE plates DROP COLUMN sparse_infill_density; ALTER TABLE plates DROP COLUMN wall_loops; ALTER TABLE default_settings DROP COLUMN sparse_infill_pattern; ALTER TABLE default_settings DROP COLUMN sparse_infill_density; ALTER TABLE default_settings DROP COLUMN wall_loops; DROP TABLE print_notifications; ALTER TABLE print_jobs DROP COLUMN estimate_json; PRAGMA user_version=7;").unwrap();
    rig.env
        .insert("NOTIFICATION_TEST_ENABLED".into(), "1".into());
    rig.launch();
    rig.report("FINISH");
    rig.phase("awaiting_removal");
    sleep(1.2);
    assert!(rows(&rig).is_empty());
    assert_eq!(webhook.len(), count);
    remove_current(&rig);
    for path in [
        "/api/queue?printer_id=p1",
        "/api/printers",
        "/api/plates",
        "/api/about",
    ] {
        assert!(!rig.get(path).to_string().contains(TOKEN));
    }
    assert!(
        !fs::read_to_string(rig.output.join("server.log"))
            .unwrap()
            .contains(TOKEN)
    );
    rig.stop(false);
}
